Rework dependencies

ottermata opened the pull request:
Dependencies currently have two problems:
1. They are resolved sequentially
2. They may be resolved multiple times within a single request

In order to solve the first without introducing locks, the resolution algorithm was changed to:
1. Recursively getting a list of all dependencies, with their required sub-dependencies
2. Creating a sequence of dependency batches. The dependencies per batch can run concurrently, but depend on values from the previous batches. Each batch is created by:
  a. Getting all unaccounted for dependencies without remaining sub-dependencies
  b. Removing those dependencies from the list of overall dependencies
  c. Removing them from the remaining sub-dependencies of the other dependencies

This way the following dependency graph will be split into these batches: `(A, B, C), (D, E), (F)`
```
┌─F─┐
│   │
▼   │
D   │   E
│   │   │
▼   ▼   ▼
A   B   C
```
This is done once when the KwargsModel is created. When actually resolving the dependencies, we only need to iterate over the batches and resolve the the dependencies without checking for sub-dependencies. In order to avoid unnecessary overhead we only create the task group when there is more than one dependency per batch.

This also solves the second problem, all dependencies are only called once per request. If in the above graph F were to also require A or C, nothing would change. If E were to require D, it would only move to F's batch.

# PR Checklist

- [X] Have you followed the guidelines in `CONTRIBUTING.md`?
- [X] Have you got 100% test coverage on new code?
- [X] Have you updated the prose documentation?
- [X] Have you updated the reference documentation?

Goldziher commented on starlite/datastructures/provide.py:
```diff
@@ -15,30 +13,24 @@
-        cache_per_request: bool = False,
-        cache_key: Optional[str] = None,
         sync_to_thread: bool = False,
+        **_kwargs: Any,
```
what are kwargs used for?

ottermata replied:
Thought this might be better than just removing the `cache_per_request` and `cache_key` options. I can remove this if that's not a concern.

Goldziher commented on starlite/kwargs.py:
```diff
@@ -409,39 +418,94 @@ def _collect_params(params: Mapping[str, Any], expected: Set[ParameterDefinition
+            connection: An instance of [Request][starlite.connection.Request] or [WebSocket][starlite.connection.WebSocket].
+            kwargs: Kwargs to pass to dependencies
+        """
+        if len(dependencies) == 0:
```
```suggestion
        if not dependencies:
```

Goldziher commented on starlite/kwargs.py:
```diff
@@ -409,39 +418,94 @@ def _collect_params(params: Mapping[str, Any], expected: Set[ParameterDefinition
+
+    async def _resolve_dependencies_concurrently(
+        self,
+        dependencies: Collection["Dependency"],
```
collection is too generic. This is not a collection, its a list or an iteratabe.

Goldziher commented on starlite/kwargs.py:
```diff
@@ -409,39 +418,94 @@ def _collect_params(params: Mapping[str, Any], expected: Set[ParameterDefinition
-        await resolve_dependencies_concurrently(self, dependency.dependencies, connection, kwargs)
         dependency_kwargs = signature_model.parse_values_from_connection_kwargs(connection=connection, **kwargs)
-        return await dependency.provide(**dependency_kwargs)
+        kwargs[dependency.key] = await dependency.provide(**dependency_kwargs)
```
dependency.provide is not always async. Its important to note this - this is configured by the user.

ottermata replied:
Not sure I understand, Provide.__call__ is async?

peterschutt commented:
It would be nice if you could provide some context to the PR. E.g., a high level description of the changes you are making to the algorithm and why (i.e., what problem is it solving?). It would make this a lot easier to review.

ottermata commented:
@peterschutt I updated the description

Goldziher commented on starlite/kwargs.py:
```diff
@@ -499,26 +508,94 @@ def to_kwargs(self, connection: Union["WebSocket", "Request"]) -> Dict[str, Any]
-            dependency: An instance of [Dependency][starlite.kwargs.Dependency]
             connection: An instance of [Request][starlite.connection.Request] or [WebSocket][starlite.connection.WebSocket].
-            **kwargs: Any kwargs to pass recursively.
+            kwargs: Kwargs to pass to dependencies
```
```suggestion
            **kwargs: Kwargs to pass to dependencies
```

This is the correct format. You can test for doc issues by running `docker compose up --build`, which will bring the docs up locally.

ottermata replied:
I don't think the ** is correct? The kwargs are passed in as a dictionary, so that they can be mutated.

Goldziher commented on starlite/kwargs.py:
```diff
@@ -499,26 +508,94 @@ def to_kwargs(self, connection: Union["WebSocket", "Request"]) -> Dict[str, Any]
 
-        Returns:
-            The resolved dependency value
+    async def _resolve_dependencies_concurrently(
```
i think most of these methods can be moved into functions from the class. There is no point in using self here.

Goldziher commented on starlite/kwargs.py:
```diff
@@ -499,26 +508,94 @@ def to_kwargs(self, connection: Union["WebSocket", "Request"]) -> Dict[str, Any]
+            kwargs: Kwargs to pass to dependencies
+        """
+        if not dependencies:
+            pass
```
this should never be the case - this function should not be called at all if dependencies is empty.

Goldziher commented on starlite/kwargs.py:
```diff
@@ -499,26 +508,94 @@ def to_kwargs(self, connection: Union["WebSocket", "Request"]) -> Dict[str, Any]
+        if not dependencies:
+            pass
+        elif len(dependencies) == 1:
+            await self._resolve_dependency(next(iter(dependencies)), connection, kwargs)
```
I dont think this should be inside thi function - it should be outside it. This function should only include whats now under `else`, and be called only when there are multiple depenendencies to resolve concurrently.

Goldziher commented on starlite/kwargs.py:
```diff
@@ -499,26 +508,94 @@ def to_kwargs(self, connection: Union["WebSocket", "Request"]) -> Dict[str, Any]
+                    task_group.start_soon(self._resolve_dependency, dependency, connection, kwargs)
+
+    @staticmethod
+    async def _resolve_dependency(
```
this shouldnt be a private static method. Instead it should be the contents of `resolve_dependency` itself, or something else (see my ending comment.

Goldziher commented on starlite/kwargs.py:
```diff
@@ -499,26 +508,94 @@ def to_kwargs(self, connection: Union["WebSocket", "Request"]) -> Dict[str, Any]
+        kwargs[dependency.key] = await dependency.provide(**dependency_kwargs)
+
+    @classmethod
+    def _get_dependency_batches(cls, expected_dependencies: Set[Dependency]) -> List[Set[Dependency]]:
```
same as previous, these two class methods should be moved out of the class. 

Regarding names:

instead of `_get_dependency_batches` --> `batch_dependencies`, `split_dependencies_into_batches`, `create_dependency_batches etc.` 

`_get_dependency_batches_helper ` --> this name is meaningless. Try ` map_dependencies_recurisvely` or something like this.

Goldziher commented on tests/dependency_injection/test_request_local_caching.py:
```diff
@@ -0,0 +1,34 @@
+from starlite.status_codes import HTTP_200_OK
+
+
+def test_caching_per_request() -> None:
```
I thought dependenc caching is gone 😉 , in fact i know its gone. 

I think you should probably rebase and remove this.

What I am not seeing-- tests that ensure the functionality you are adding works as expected. And I would like to run benchmarks to ensure that this improves performance and doesn't cause memory issues.

ottermata replied:
Dependency "caching" gets reintroduced (every dependency is only resolved once), so I reused the test. Maybe change the name to be more clear?

I added a test for the batching.

Goldziher reviewed (commented):
Interesting contribution! I would like to be a bit more cautious with this. i am not sure what the performance gain of using a task group here gives us - it adds complexity and we need to ascertain this. I am sure @provinzkraut can help us with testing DI in some way. 

The other comment I have is more general - if you look at the logic in `kwargs.py` on main, you will see that i added `extractors` there. There is a function called `create_connection_value_extractor`, which is used to creat a precomputed extractor for data for different types. I think we should opt in to precompute a function that handles dependencies for a given kwargs model in advance, not repeat the calculation on each request.

ottermata commented:
> Interesting contribution! I would like to be a bit more cautious with this. i am not sure what the performance gain of using a task group here gives us - it adds complexity and we need to ascertain this. I am sure @provinzkraut can help us with testing DI in some way.

Feel free to test the performance of this, I rebased against main.

> The other comment I have is more general - if you look at the logic in `kwargs.py` on main, you will see that i added `extractors` there. There is a function called `create_connection_value_extractor`, which is used to creat a precomputed extractor for data for different types. I think we should opt in to precompute a function that handles dependencies for a given kwargs model in advance, not repeat the calculation on each request.

The dependency batches are stored on the KwargsModel, just like the extractors. So I assume they should be cached just the same? I refactored the code to create the batches within the constructor, makes it a lot simpler as well.
